average.js

```diff
@@ -1,6 +1,7 @@
 
 function average(numbers) {
-  return numbers.reduce((p, c)=> p + c, 0) / numbers.length;
+  const filteredNumbers = numbers.filter((number) => !isNaN(number));
+  return filteredNumbers.reduce((p, c)=> p + c, 0) / filteredNumbers.length;
 }
 
-module.exports = {average};
+module.exports = {average};
```

Comment:
Try a new test: The tests in `test/average.test.js` express the functionality of the `average` function.
The functionality accepts a sequence of readings from an IoT sensor. The sensor sends `NaN` when something goes wrong in measurement, which we have ignored.
Problem: If the sensor sends a lot of NaNs and very few readings, something is wrong. However, the `average` function ignores all the NaNs and still computes the average, as if nothing went wrong.
We need to add functionality to the average function to return `NaN`, when most of the input-numbers are `NaN`.
You need to:

- Reason out what 'most of the numbers' means: E.g., determine the number of NaNs in the input which would indicate that the sensor is unreliable. Mention any references you used to arrive at this logic
- Add a test that supplies such inputs and expects the output to be NaN
- Add code that satisfies this functionality, while not breaking existing functionality.

Reply:
Hello, I have tried to implement what you asked for.
I tried a lot for finding some references suggesting some kind of threshold or something for sensors but I couldn't find any direct references indicating the same.
I am attaching the links which I referred to arrive to my solution.

1. [https://discuss.analyticsvidhya.com/t/what-should-be-the-allowed-percentage-of-missing-values/2456](https://discuss.analyticsvidhya.com/t/what-should-be-the-allowed-percentage-of-missing-values/2456)
2. [https://www.researchgate.net/publication/322067050_IMPUTING_THE_MISSING_VALUES_IN_IOT_USING_ESTCP_MODEL](https://www.researchgate.net/publication/322067050_IMPUTING_THE_MISSING_VALUES_IN_IOT_USING_ESTCP_MODEL)
3. [https://towardsdatascience.com/preprocessing-iot-data-linear-resampling-dde750910531](https://towardsdatascience.com/preprocessing-iot-data-linear-resampling-dde750910531)

I have considered two cases:
**Case 1: Percentage of NaN readings in the input.**
So from the above links and some other papers which i found suggested that theoritically 25-30% is the maximum null/missing data values should be allowed.
Also found some papers where in they have tried to interpolate the missing data which showed results that if more than 25% data is missing then its difficult to predict the missing values.
Also the towardsdatascience blog showed something similar, well not exactly in terms of percentage but it says that if sensors remain offline / faulty for more than 1 hour then dealing/predicting the missing data becomes difficult.

So i have kept the threshold to 30% i.e. if percentage of NaN reading exceeds 30% then we return NaN.

**Case 2: Consecutive NaN readings in the input.**
This case i have included on my own since i didn't find any direct reference for percentage NaN readings upto which we should ignore.
So there might be some cases where the servers are down for some time sending continuous NaN values and some other factors as well.

So If the length of consecutive NaN values in input is greater than THRESHOLD then we return NaN.
THRESHOLD which I have defined for this is square root of input length. This is because I felt the count should be less than the above case i.e. 30% of input length. So in case of larger size inputs even if there is only 20% NaN values in the input but if all those are in consecutive sequence then we should handle that case and return NaN indicating faulty sensor.


Please review and let me know of any further task which you want me to do @numocityadmin @sudeeprp .